### ISSUE TITLE
Stash function sometimes not working in 2.2.1

### ISSUE THREAD

**Issue by rbrahmson** [vanilla/vanilla#4391]
I have a plugin (DiscussionTopic) that uses the stash function for two functionalities:

1. To store a message that is later displayed in the discussion list
2. To store a string that is later used in DiscussionModel_BeforeGet_Handler as an argument to "$Sender->SQL->Where('d.Topic',$Search);" ($Search is the filtering string).

The first option works as intended.  The second one (in the BeforeGet hook) causes a failure.

I have since bypassed the issue in a new version of the plugin by using cookies instead of the stash function.  The source of the original problematic version is in a the Github history [here](https://github.com/rbrahmson/VanillaPlugins/blob/f3273b6b4e4d636fef738a5e177bc820e25c86db/DiscussionTopic/class.discussiontopic.plugin.php) line 1726.

I include an error screen sent by one of the Vanilla forum users who tried the plugin (he is on 2.2.).

P.S. The version with the error does not conform to the Vanilla coding standard in any way.  The current version in Github (with the cookie-method bypass) has fixed many of the standard coding issues (few more to be fixed in the future) and is therefore more readable.

![stash error](https://cloud.githubusercontent.com/assets/6851541/18204342/3fcd1400-70ea-11e6-86c0-959ee1bc4cb0.png)

**Comment by R-J** [comment]
stash() uses Gdn::sql() to work with the Session table.  Discussion model also uses Gdn::sql(). That way, both work on the same object and when the event is fired in the model, some parts of a query have already been built. This unfinished query gets mixed with the query that is built in the stash() method which obviously has to fail.

@rbrahmson:  whenever you hook into a db event and you want to do any db interactions, this is something to look at. See class UserMetaModels __construct method how this can be avoided.

Since the class Session only has two methods which use the db, adding something like that in the constructor might be too much. Instead using a local clone could be better. I'll make a quick PR...

**Comment by rbrahmson** [comment]
Good advice, but I think that the stash function should be using a new sql object to shield it a SQL query being partially built for a different purpose.  Looking forward to see and learn from your PR.

**Comment by R-J** [comment]
No need for the "but" -  I also think it shouldn't use Gdn:sql(). Since every finished query deletes everything that has been added to the query builder before, there might be some unpredictable misbehaviors.  Imagine you have

~~~
// Do some pagination
// Process RequestArgs for page number
// Calculate limit and offset
// Apply them to the query which should be made
Gdn::sql()->limit($offset, $limit);

// Now prepare some other query parts
$value = Gdn::session->stash($key);
Gdn::sql->select();
Gdn::sql->where($column, $value);
// ...
~~~

That would result in a (most probably) correct result of the stash call, but whatever should be querried from db wouldn't be what was expected and the reason is buried in the Session class.

Obviously that will be changed one way or the other.

**Comment by rbrahmson** [comment]
Linc - in light of your post in VF that 2.3 is almost ready, I'm reminding you that this issue is still open here.  You seemed to accept RJ's suggestion but your response above was not understood by either R_J or myself.  Can you please confirm the fix for 2.3 and clarify?

**Comment by rbrahmson** [comment]
Thanks!
